phper/src/builder/function.rs

```diff
@@ -0,0 +1,101 @@
+use crate::builder::argument::Argument;
+use crate::builder::argument::ArgumentSet;
+use crate::builder::data_type::AtomicReturnType;
+use crate::builder::data_type::AtomicType;
+use crate::builder::describe::Describe;
+use crate::builder::handler::Context;
+use crate::builder::handler::FunctionHandler;
+use crate::builder::handler::State;
+use crate::builder::value::Value;
+use crate::errors::Result;
+
+pub struct Function<F>
+where
+    F: Fn(State, Context, ArgumentSet) -> Result<Value>,
+{
+    pub name: String,
+    pub arguments: Vec<Argument>,
+    pub returns: AtomicReturnType,
+    pub handler: FunctionHandler<F>,
+}
+
+impl<F> Function<F>
+where
+    F: Fn(State, Context, ArgumentSet) -> Result<Value>,
+{
+    pub fn new(name: String, handler: FunctionHandler<F>) -> Function<F> {
+        Function {
+            name,
+            arguments: vec![],
+            returns: AtomicReturnType::Of(AtomicType::Mixed),
+            handler,
+        }
+    }
+
+    pub fn with_argument(mut self, argument: Argument) -> Self {
+        self.arguments.push(argument);
+
+        self
+    }
+
+    pub fn returns(mut self, return_type: AtomicReturnType) -> Self {
+        self.returns = return_type;
+
+        self
+    }
+}
+
+/// Describe `Function`.
+///
+/// Example:
+///
+/// ```rust
+/// use phper::builder::argument::Argument;
+/// use phper::builder::data_type::AtomicReturnType;
+/// use phper::builder::data_type::AtomicType;
+/// use phper::builder::describe::Describe;
+/// use phper::builder::function::Function;
+/// use phper::builder::handler::Context;
+/// use phper::builder::handler::FunctionHandler;
+/// use phper::builder::handler::State;
+/// use phper::builder::value::ConstantValue;
+/// use phper::builder::value::Value;
+///
+/// let handler = FunctionHandler(|_, _, arguments| {
+///     let mut message: String = arguments.get("message")?.value()?.try_into()?;
+///     let args: Vec<String> = arguments.get("args")?.value()?.try_into()?;
+///     for (i, arg) in args.iter().enumerate() {
+///         // replace `{n}` with the args[n]
+///         message = message.replace(&format!("[{i}]"), arg);
+///     }
+///
+///     Ok(ConstantValue::String(message).into())
+/// });
+///
+/// let function = Function::new("format".to_owned(), handler)
+///     .with_argument(Argument::new("message", AtomicType::String))
+///     .with_argument(Argument::varadic("args", AtomicType::String))
+///     .returns(AtomicReturnType::Of(AtomicType::String));
+///
+/// assert_eq!(
+///     "function format(string $message, string ...$args): string { exit(1); }",
+///     function.describe()
+/// );
+/// ```
```

Comment:
@jmjoy hopefully this gives you an idea for what I'm going for, i would like to get your opinion on it and the rest of the API before i continue :)